File: desktop/lib/src/desktop/theme/status/circular_progress_indicator.dart
import 'package:flutter/widgets.dart';

import '../color_scheme.dart';

part 'circular_progress_indicator.g.dart';

/// Theme data for [CircularProgressIndicator].
@immutable
class _CircularProgressIndicatorThemeData {
  const _CircularProgressIndicatorThemeData(ThemeData themeData)
    : _themeData = themeData;

  final ThemeData _themeData;

  ColorScheme get _colorScheme => _themeData.colorScheme;

  /// The size of the circular progress indicator.
  ///
  /// Defaults to:
  ///
  /// ```dart
  /// 40.0
  /// ```
  double get size => 24.0;

  /// The stroke width of the circular progress indicator.
  ///
  /// Defaults to:
  ///
  /// ```dart
  /// 2.0
  /// ```
  double get strokeWidth => 2.0;

  /// The color of the circular progress indicator. Defaults to []
  ///
  /// Defaults to:
  ///
  /// ```dart
  /// colorScheme.primary[50]
  /// ```
  Color get color => _colorScheme.primary[50];

  /// The background color of the circular progress indicator.
  ///
  /// Defaults to:
  ///
  /// ```dart
  /// colorScheme.disabled
  /// ```
  Color get backgroundColor => _colorScheme.disabled;

  /// The indeterminate animation duration.
  ///
  /// Defaults to:
  ///
  /// ```dart
  /// Duration(milliseconds: 6400)
  /// ```
  Duration get indeterminateDuration => const Duration(milliseconds: 3000);
}

File: desktop_swift/Sources/Desktop/Status/CircularProgressIndicator.swift
import Foundation

#if canImport(SwiftUI)
import SwiftUI

/// A themed circular progress indicator, mirroring the Flutter
/// `CircularProgressIndicator`.
///
/// If `value` is `nil`, the indicator spins continuously.
public struct CircularProgressIndicator: View {
    @Environment(\.desktopTheme) private var theme

    private let value: Double?
    private let size: CGFloat
    private let lineWidth: CGFloat

    @State private var rotation: Double = 0

    public init(value: Double? = nil, size: CGFloat = 24, lineWidth: CGFloat = 3) {
        self.value = value
        self.size = size
        self.lineWidth = lineWidth
    }

    public var body: some View {
        ZStack {
            Circle()
                .stroke(theme.colorScheme.background[12], lineWidth: lineWidth)
            if let value = value {
                Circle()
                    .trim(from: 0, to: max(0, min(1, value)))
                    .stroke(theme.colorScheme.primary[50],
                            style: StrokeStyle(lineWidth: lineWidth, lineCap: .round))
                    .rotationEffect(.degrees(-90))
            } else {
                Circle()
                    .trim(from: 0, to: 0.25)
                    .stroke(theme.colorScheme.primary[50],
                            style: StrokeStyle(lineWidth: lineWidth, lineCap: .round))
                    .rotationEffect(.degrees(rotation))
                    .onAppear {
                        withAnimation(.linear(duration: 1).repeatForever(autoreverses: false)) {
                            rotation = 360
                        }
                    }
            }
        }
        .frame(width: size, height: size)
    }
}
#endif
